zebra/zebra_ted.c

```diff
@@ -0,0 +1,280 @@
+	stream_fifo_init(&zt_info.in_fifo);
+
+	zt_info.msgs_per_cycle = ZEBRA_TED_MSG_LIMIT;
+}
+
+/*
+ * Start the module pthread. This step is run later than the
+ * 'init' step, in case zebra has fork-ed.
+ */
+void zebra_ted_start(void)
+{
+	struct frr_pthread_attr pattr = {
+		.start = frr_pthread_attr_default.start,
+		.stop = frr_pthread_attr_default.stop
+	};
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module starting", LOG_NAME);
+
+	/* Start pthread */
+	zt_info.pthread = frr_pthread_new(&pattr, "Zebra TED thread",
+					  "zebra_TED");
+
+	/* Associate event 'master' */
+	zt_info.master = zt_info.pthread->master;
+
+	zt_info.run = true;
+
+	/* Enqueue an initial event for the pthread */
+	thread_add_event(zt_info.master, process_messages, NULL, 0,
+			 &zt_info.t_msgs);
+
+	/* And start the pthread */
+	frr_pthread_run(zt_info.pthread, NULL);
+}
+
+/*
+ * Module stop, halting the dedicated pthread; called from the main pthread.
+ */
+void zebra_ted_stop(void)
+{
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module stop", LOG_NAME);
+
+	zt_info.run = false;
+
+	frr_pthread_stop(zt_info.pthread, NULL);
+
+	THREAD_OFF(zt_info.t_msgs);
+
+	frr_pthread_destroy(zt_info.pthread);
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module stop complete", LOG_NAME);
+}
+
+/*
+ * Module final cleanup, called from the zebra main pthread.
+ */
+void zebra_ted_finish(void)
+{
+	struct stream *msg;
+
+	if (IS_ZEBRA_DEBUG_EVENT)
+		zlog_debug("%s module shutdown", LOG_NAME);
+
+	/* Clean up any queued messages */
+	msg = stream_fifo_pop(&zt_info.in_fifo);
+	while (msg) {
+		stream_free(msg);
+		msg = stream_fifo_pop(&zt_info.in_fifo);
+	}
+}
+
+/*
+ * Does this module handle (intercept) the specified zapi message type?
+ */
+bool zebra_ted_handle_msgid(uint16_t id)
+{
+	bool ret = false;
+
+	switch (id) {
+	case ZEBRA_OPAQUE_MESSAGE:
+		ret = true;
+		break;
+	default:
+		break;
+	}
+
+	return ret;
+}
+
+/*
+ * Enqueue a batch of messages for processing - this is the public api
+ * used from the zapi processing threads.
+ */
+uint32_t zebra_ted_enqueue_batch(struct stream_fifo *batch)
+{
+	uint32_t counter = 0;
+	struct stream *msg;
+
+	/* Dequeue messages from the incoming batch, and save them
+	 * on the module fifo.
+	 */
+	frr_with_mutex(&zt_info.mutex) {
+		msg = stream_fifo_pop(batch);
+		while (msg) {
+			stream_fifo_push(&zt_info.in_fifo, msg);
+			counter++;
+		}
+	}
+
+	/* Schedule module pthread to process the batch */
+	if (counter > 0) {
+		if (IS_ZEBRA_DEBUG_EVENT)
+			zlog_debug("%s: received %u messages",
+				   __func__, counter);
+		thread_add_event(zt_info.master, process_messages, NULL, 0,
+				 &zt_info.t_msgs);
+	}
+
+	return counter;
+}
+
+
+/*
+ * Pthread event loop, process the incoming message queue.
+ */
+static int process_messages(struct thread *event)
+{
+	struct stream_fifo fifo;
+	struct stream *msg;
+	uint32_t i;
+	bool need_resched = false;
+
+	/* Check for zebra shutdown */
+	if (!zt_info.run)
```

Comment:
It would be better to check the `running` field in the `frr_pthread` struct itself, which has the same information. That check will need to be atomic as well.